File: SaloniDattani/Life-expectancy/Mortality-rates-over-time-countries/inputs/mortality-rates-over-time-countries.R
# Open libraries
library(tidyverse)
library(scales)
library(viridis)

# Data source:
# https://www.mortality.org/
# Choose countries, then go to Cohort data > Death Rates > 1x10
# Download and replace this with path to folder
data_folder <- ""
# set filename to cMx_1x10_(name of country).txt

countries <- c("USA", "Italy", "UK", "Sweden")
mortality <- list()

# Import data
for (country in countries) {
  
  # Import and rename cols
  mortality[[country]] <- read_table(paste0(data_folder, "cMx_1x10_", country, ".txt"), skip=2)
  colnames(mortality[[country]]) <- c("Year", "Age", "Female", "Male", "Total")
  
  mortality[[country]] <- mortality[[country]] %>%
    mutate(Country = country)
  
}

# Join into single df
mortality <- do.call(rbind.data.frame, mortality)

# Reformat
mortality$Age <- as.integer(mortality$Age)
mortality$Female <- as.numeric(mortality$Female)
mortality$Male <- as.numeric(mortality$Male)
mortality$Total <- as.numeric(mortality$Total)
mortality$Country <- as.factor(mortality$Country)

# Gather to make it long format and get only values for total
mortality_g <- gather(mortality, "Demographic", "Rate", 3:5)
mortality_g_total <- filter(mortality_g, Demographic == "Total")

# Select decades to show
decades <- seq(1800, 2000, by=10)

mortality_g_total <- mortality_g_total %>%
                      filter(Year %in% paste0(decades, "-", decades + 9))

# Get number of time periods shown for colour scale
n_colours <- length(decades)
colors <- rev(turbo(n_colours))

# Plot
ggplot(data=mortality_g_total, aes(color=Year, x=Age, y=Rate)) +
  # Choose line or smoothed line or points
  geom_line(aes(color=Year),size=1,alpha=1) +
  geom_point(data=filter(mortality_g_total, Age==0), aes(color=Year,x=Age,y=Rate), size=1, show.legend=FALSE)+
  # Limit to 95 because ages above 100 are noisy and go above 100%
  coord_cartesian(xlim=c(0,95)) +
  facet_grid(cols=vars(Country)) +
  theme_classic() +
  theme(strip.background = element_blank()) +
  scale_x_continuous(breaks = seq(0, 100, by=10)) +
  scale_y_continuous(labels = scales::percent, trans='log2', breaks = c(0.0001, 0.001, 0.01, 0.1, 1)) +
  scale_color_manual(name = "Birth cohort", values = colors) +
  labs(title = "Annual death rate by age", 
       y = "Death rate", 
       x = "Age",
       color = "Birth cohort",
       caption = "Cohort death rates.\nSource: Human Mortality Database.\nUniversity of California, Berkeley (USA)\n(data downloaded on [12 Sep 2023])") 

ggsave(paste0(data_folder, "annual-mortality-time-countries.svg"), width = 12, height = 6)

File: SaloniDattani/Life-expectancy/Mortality-rates-over-time-countries/inputs/mortality-rates-over-time-countries.py
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

# Data source:
# https://www.mortality.org/
# Replace this with path to folder where you saved it
data_folder = ""
# Replace with names of countries
countries = ["Italy", "Sweden"]
mortality = {}

# Import data
for country in countries:
    # Import and rename cols
    path = f"{data_folder}cMx_1x10_{country}.txt"
    mortality[country] = pd.read_csv(path, delim_whitespace=True, skiprows=2, names=["Year", "Age", "Female", "Male", "Total"])
    mortality[country]['Country'] = country

# Join into single df
mortality_df = pd.concat(mortality.values())

# Convert to correct datatypes
mortality_df['Age'] = pd.to_numeric(mortality_df['Age'], errors='coerce')
mortality_df['Year'] = mortality_df['Year'].astype(str)
mortality_df = pd.melt(mortality_df, id_vars=['Year', 'Age', 'Country'], value_vars=['Female', 'Male', 'Total'], var_name='Demographic', value_name='Rate')
mortality_g_total = mortality_df[mortality_df['Demographic'] == "Total"]

# Log transformation of Rate for plotting
mortality_g_total['Rate'] = pd.to_numeric(mortality_g_total['Rate'], errors='coerce')
mortality_g_total = mortality_g_total.dropna(subset=['Rate'])
mortality_g_total = mortality_g_total[mortality_g_total['Rate'] > 0].reset_index(drop=True)
mortality_g_total['Log_Rate'] = np.log2(mortality_g_total['Rate'])


plt.figure(figsize=(12, 6))

# Create a color palette
n_colors = mortality_g_total['Year'].nunique()
colors = sns.color_palette("viridis", n_colors)
color_dict = dict(zip(mortality_g_total['Year'].unique(), colors))

# Line plot for each year's data
sns.lineplot(data=mortality_g_total, x='Age', y='Rate', hue='Year', palette=color_dict, alpha=0.5)

# Scatter plot for age 0 points
age_zero_data = mortality_g_total[mortality_g_total['Age'] == 0]
sns.scatterplot(data=age_zero_data, x='Age', y='Rate', hue='Year', palette=color_dict, legend=False, edgecolor='black')

# Set the y-axis scale to log2 and customize the appearance
plt.yscale('log', base=2)
plt.yticks([0.0001, 0.001, 0.01, 0.1, 1], ['0.01%', '0.1%', '1%', '10%', '100%'])
plt.ylim(0.0001, 1)

# Add title and labels
plt.title("Annual death rate by age")
plt.xlabel("Age")
plt.ylabel("Death rate (log2 scale)")

# Add grid and facet by Country
g = sns.FacetGrid(mortality_g_total, row="Country", margin_titles=True)
g.map_dataframe(sns.lineplot, x='Age', y='Rate', hue='Year', palette=color_dict, alpha=0.5)
g.map_dataframe(sns.scatterplot, data=age_zero_data, x='Age', y='Rate', hue='Year', palette=color_dict, edgecolor='black', legend=False)
g.set(yscale="log", base=2)
g.set(yticks=[0.0001, 0.001, 0.01, 0.1, 1], yticklabels=['0.01%', '0.1%', '1%', '10%', '100%'])

plt.show()

